client/lib/track-element-size/index.ts

```diff
@@ -0,0 +1,118 @@
+/**
+ * External dependencies
+ */
+import { useRef, useEffect, useState, useCallback } from 'react';
+import { throttle } from 'lodash';
+
+/**
+ * Internal dependencies
+ */
+import afterLayoutFlush from 'lib/after-layout-flush';
+
+const THROTTLE_RATE = 200;
+
+// Indexable version of DOMRect / ClientRect.
+interface UseWindowResizeRect {
+	[index: string]: number | undefined;
```

Comment:
I'm not entirely sure whether this syntax allows for additional properties. That was not the intention; rather, I was looking for a way of indicating that the properties are indexable through `Object.keys`. Is there a better way?

Reply:
I made the checks explicit (instead of using a loop), in order to avoid this issue in the first place.

Reply:
What was the issue in the first place? On my machine, TypeScript doesn't seem to mind if I pass any valid object to `Object.keys`. It doesn't require any particular subtype.

Reply:
The issue is array-type indexing:

```ts
interface Test {
  foo: string,
  bar: string
}

const test: Test = { foo: 'a', bar: 'b' };

// This is fine.
const array1 = [ test.foo, test.bar ];
// This is not, because of the `[prop]`, even though those are valid prop values.
const array2 = [ 'foo', 'bar' ].map( prop => test[ prop ] );
```

Reply:
The problem is that `Object.keys` returns `string[]`. TypeScript complains when we try to access properties of a defined interface with an index of type `string` because the key type is a union of the known keys `"top" | "bottom" | …`.

One option is to open up the interface so it accepts arbitrary string keys, the other is to be more strict.

We know the keys of the object will be the keys in the interface, but I think the way TypeScript works makes it impossible to make those types work.

An alternative is to tell TypeScript:

```ts
interface I {  a: string; }

const anObject = { a: "a" };
const anI: I = { a: "a" };

// Not happy with this because `Object.keys` returns `string[]`:
//
// for (const k of Object.keys(anI)) {
//
// If we say it's an array of the keys of the interface, we're all good: 
for (const k of Object.keys(anI) as (keyof I)[]) {
    if ( anI[k] === anObject[k] ) {
        console.log( 'equal' );
    }
}
```

Reply:
> // This is not, because of the `[prop]`, even though those are valid prop values.
> const array2 = [ 'foo', 'bar' ].map( prop => test[ prop ] );

This is the same issue, `[ 'foo', 'bar']` is typed as `string[]`, but if we help out like this, it's OK:

```ts
const array2 = (["foo", "bar"] as (keyof Test)[]).map(prop => test[prop]);
```

Reply:
TIL about `keyof`! This looks like an interesting way of handling similar situations in the future. Thanks, @sirreal !

Reply:
I see, I can't just access `obj[ prop ]`, because TypeScript doesn't know the value of `prop` and can't typecheck it. Can be annoying, but makes sense.